### PR TITLE
feat(proposals): add nested API objects

### PR DESCRIPTION
**Pull request by jbrockopp**
Co-authored-by: @ecrupper

Please vote for which option you prefer the most:

* [Option 1](https://github.com/go-vela/community/blob/feature/proposal/nested_api_objects/proposals/2022/07-01_nested_api_objects.md#option-1)
* [Option 2](https://github.com/go-vela/community/blob/feature/proposal/nested_api_objects/proposals/2022/07-01_nested_api_objects.md#option-2)

### CODE REVIEW

File: proposals/2022/07-01_nested_api_objects.md [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+Using the above endpoint (`GET /api/v1/repo/:org/:repo`), the result would replace the `user_id` field with an `owner` field:
+
+```diff
+{
+  "id": 1,
+- "user_id": 1,
++ "owner": {
++   "id": 1,
++   "name": "OctoKitty",
++   "favorites": ["github/octocat"],
```
Comment (cognifloyd) [inline proposals/2022/07-01_nested_api_objects.md:144]: I don't think favorites should be exposed in most cases. It seems to me that that's a user preference that the UI (or CLI) should be able to request for the current user, bit nobody/nothing else needs it.

What purpose does exposing a user's favorites to a build (in env vars) or in these API calls have? Is there a use case I haven't considered?

**Review by cognifloyd — COMMENTED** [review]
I prefer option 1 as Vela has not released v1 yet, so, based on semantic versioning, API changes should be expected.

File: proposals/2022/07-01_nested_api_objects.md (hunk 1 above)
Comment (jbrockopp) [reply proposals/2022/07-01_nested_api_objects.md:144]: I don't think we necessarily meant to propose that the `owner` field has to have the `favorites` displayed.

More so, I think we simply took the [existing user API definition](https://github.com/go-vela/types/blob/master/library/user.go#L13-L25):

```go
// User is the library representation of a user.
//
// swagger:model User
type User struct {
	ID           *int64    `json:"id,omitempty"`
	Name         *string   `json:"name,omitempty"`
	RefreshToken *string   `json:"-"`
	Token        *string   `json:"-"`
	Hash         *string   `json:"-"`
	Favorites    *[]string `json:"favorites,omitempty"`
	Active       *bool     `json:"active,omitempty"`
	Admin        *bool     `json:"admin,omitempty"`
}
```

And removed the fields that can't be returned (`-` in JSON field tags).

Regarding implementation, we can certainly ensure that `favorites` is not returned to minimize payload size.

**Comment by cognifloyd** [comment]
Here's an interesting approach to API compatibility guarantees:
Vault clarifies that `/v1` is not finalized yet in the docs:
> **Backwards compatibility**: At the current version, Vault does not yet promise backwards compatibility even with the `v1` prefix. We'll remove this warning when this policy changes. At this point in time the core API (that is, `sys/` routes) change very infrequently, but various secrets engines/auth methods/etc. sometimes have minor changes to accommodate new features as they're developed.

We could copy that approach and add a note to the docs about compatibility.

**Comment by chrispdriscoll** [comment]
Per monthly review of open PRs, the Vela Admins will keep this proposal open until PRs are raised.

**Review by plyr4 — APPROVED** [review]

**Review by KellyMerrick — APPROVED** [review]
lgtm

**Pull request merged**